plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/TypeJudgment.java

```diff
@@ -607,12 +612,55 @@ public TypeRef caseIdentifierRef(IdentifierRef idref) {
 							}
 
 							TypeUtils.sanitizeRawTypeRef(instanceofType);
-							intersectionTypes.add(instanceofType);
+							if (ioGuard.asserts == GuardAssertion.AlwaysHolds) {
+								allTypes.add(instanceofType);
+							} else if (ioGuard.asserts == GuardAssertion.NeverHolds) {
+								excludedTypes.add(instanceofType);
+							}
 						}
 					}
 				}
-				if (!intersectionTypes.isEmpty()) {
-					T = tsh.createIntersectionType(G, intersectionTypes.toArray(new TypeRef[intersectionTypes.size()]));
+
+				if (excludedTypes.isEmpty()) {
+					remainingTypes.addAll(allTypes);
+				} else {
+					for (TypeRef origType : allTypes) {
+						TypeRef remainingType = null;
+						if (origType instanceof UnionTypeExpression) {
+							final UnionTypeExpression union = (UnionTypeExpression) origType;
+							final Collection<TypeRef> newUnionTypes = new LinkedList<>();
+							for (TypeRef unionType : union.getTypeRefs()) {
+								for (TypeRef neverHoldingType : excludedTypes) {
+									if (!ts.equaltypeSucceeded(G, unionType, neverHoldingType)) {
+										newUnionTypes.add(unionType);
+									}
+								}
```

Comment:
Only in case there are multiple instanceof type guards that have a right hand side which evaluates to the same type. Note also the simplification logic of the intersection creation.